core/workspace_svg.js

```diff
@@ -901,14 +901,30 @@ Blockly.WorkspaceSvg.prototype.paste = function(xmlBlock) {
   Blockly.Events.disable();
   try {
     var block = Blockly.Xml.domToBlock(xmlBlock, this);
-    // Rerender to get around problem with IE and Edge not measuring text
-    // correctly when it is hidden.
-    if (goog.userAgent.IE || goog.userAgent.EDGE) {
-      var blocks = block.getDescendants();
-      for (var i = blocks.length - 1; i >= 0; i--) {
-        blocks[i].render(false);
+
+    var blocks = block.getDescendants();
+    for (var i = blocks.length - 1; i >= 0; i--) {
+      var descendant = blocks[i];
+
+      // Scratch-specific: Give shadow dom new IDs to prevent duplicating on paste
+      for(var j = 0; j < descendant.inputList.length; j++) {
+        if (descendant.inputList[j].connection) {
+          var connection = descendant.inputList[j].connection;
+          if (connection && connection.getShadowDom()) {
+            var shadowDom = connection.getShadowDom();
```

Comment:
Could be simplified a little bit since we already know if the connection exists
```javascript
        if (descendant.inputList[j].connection) {
          var connection = descendant.inputList[j].connection;
          var shadowDom = connection.getShadowDom();
          if (shadowDom) { ...
```